File: modules/purple_team.py
"""
purple_team.py — Adversary Emulation & MITRE ATT&CK Mapper
IMPROVEMENTS:
  - Works with updated database.py dict return format
  - MITRE report can be exported to a Markdown file
  - Dynamic mapping: unknown modules get a best-effort guess from keywords
  - Detection rules rendered as proper Rich syntax panels
  - Added ATT&CK Navigator layer export (JSON) for importing into the web tool
  - Graceful handling of empty DB
"""

import json
import datetime
import questionary
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from rich.syntax import Syntax
from core.ui import draw_header, Q_STYLE
from core.database import db

console = Console()

# ─────────────────────────────────────────────────────────────────────────────
#  STATIC MITRE MAPPINGS  (module name → ATT&CK data)
# ─────────────────────────────────────────────────────────────────────────────

MITRE_MAPPINGS: dict[str, dict] = {
    "Nmap-Engine": {
        "ttp":      "T1046",
        "tactic":   "Discovery",
        "name":     "Network Service Discovery",
        "splunk":   'index=network sourcetype=firewall action=allowed\n| stats count by src_ip, dest_port\n| where count > 100',
        "sigma": (
            "title: High Volume Port Scan Detected\n"
            "logsource:\n  category: firewall\ndetection:\n"
            "  selection:\n    dest_port|count: '>50'\n  condition: selection"
        ),
    },
    "AD Engine": {
        "ttp":      "T1558.004",
        "tactic":   "Credential Access",
        "name":     "AS-REP Roasting",
        "splunk":   'index=wineventlog EventCode=4768 TicketOptions=0x40810010\n| stats count by src_ip, Account_Name',
        "sigma": (
            "title: AS-REP Roasting (Event 4768)\n"
            "logsource:\n  product: windows\n  service: security\n"
            "detection:\n  selection:\n    EventID: 4768\n    TicketOptions: '0x40810010'\n  condition: selection"
        ),
    },
    "Live Interceptor": {
        "ttp":      "T1040",
        "tactic":   "Credential Access",
        "name":     "Network Sniffing",
        "splunk":   'index=sysmon EventCode=1\n| where match(Image, "wireshark|tcpdump|tshark")\n| stats count by ComputerName, Image',
        "sigma": (
            "title: Network Sniffer Launched\n"
            "logsource:\n  category: process_creation\ndetection:\n"
            "  selection:\n    Image|contains:\n      - 'tcpdump'\n      - 'wireshark'\n      - 'tshark'\n  condition: selection"
        ),
    },
    "MITM-DNS": {
        "ttp":      "T1557.002",
        "tactic":   "Collection",
        "name":     "ARP Cache Poisoning",
        "splunk":   'index=network sourcetype=arp\n| stats count by src_mac, src_ip\n| where count > 50',
        "sigma": (
            "title: ARP Poisoning Detected\n"
            "logsource:\n  category: network\ndetection:\n"
            "  selection:\n    arp_type: reply\n  condition: selection | count() > 50"
        ),
    },
    "MITM-Payload": {
        "ttp":      "T1557.001",
        "tactic":   "Collection",
        "name":     "LLMNR/NBT-NS Poisoning and SMB Relay",
        "splunk":   'index=network sourcetype=dns\n| stats count by src_ip, query\n| where count > 100',
        "sigma": (
            "title: Suspicious DNS Query Volume\n"
            "logsource:\n  category: dns\ndetection:\n"
            "  selection:\n    dns_type: query\n  condition: selection | count() > 100"
        ),
    },
    "Cloud-Ops": {
        "ttp":      "T1552.005",
        "tactic":   "Credential Access",
        "name":     "Cloud Instance Metadata API",
        "splunk":   'index=aws sourcetype=aws:cloudtrail\n| where EventName="GetSessionToken" OR dest_ip="169.254.169.254"\n| stats count by src_ip, EventName',
        "sigma": (
            "title: Suspicious Cloud Metadata Query\n"
            "logsource:\n  category: network\ndetection:\n"
            "  selection:\n    dest_ip: '169.254.169.254'\n  condition: selection"
        ),
    },
    "MSF-Engine": {
        "ttp":      "T1190",
        "tactic":   "Initial Access",
        "name":     "Exploit Public-Facing Application",
        "splunk":   'index=ids sourcetype=suricata\n| search alert.signature="*Metasploit*" OR alert.signature="*exploit*"\n| stats count by src_ip, dest_ip, alert.signature',
        "sigma": (
            "title: Remote Code Execution Payload Dropped\n"
            "logsource:\n  category: process_creation\ndetection:\n"
            "  selection:\n    ParentImage|endswith: 'w3wp.exe'\n    Image|endswith: 'cmd.exe'\n  condition: selection"
        ),
    },
    "PrivEsc-Looter": {
        "ttp":      "T1082",
        "tactic":   "Discovery",
        "name":     "System Information Discovery",
        "splunk":   'index=sysmon EventCode=1\n| where match(CommandLine, "whoami|uname|systeminfo|id")\n| stats count by ComputerName, CommandLine',
        "sigma": (
            "title: Automated PrivEsc Discovery Commands\n"
            "logsource:\n  category: process_creation\ndetection:\n"
            "  selection:\n    CommandLine|contains|all:\n      - 'whoami'\n      - '/priv'\n  condition: selection"
        ),
    },
    "Campaign-Scanner": {
        "ttp":      "T1595",
        "tactic":   "Reconnaissance",
        "name":     "Active Scanning",
        "splunk":   'index=network sourcetype=firewall\n| stats dc(dest_port) as port_count by src_ip\n| where port_count > 20',
        "sigma": (
            "title: Active Network Scanning Detected\n"
            "logsource:\n  category: firewall\ndetection:\n"
            "  selection:\n    action: deny\n  condition: selection | count(dest_port) > 20"
        ),
    },
    "Burp-Proxy": {
        "ttp":      "T1539",
        "tactic":   "Credential Access",
        "name":     "Steal Web Session Cookie",
        "splunk":   'index=web sourcetype=access_combined\n| search uri="/login" method=POST\n| stats count by clientip, uri',
        "sigma": (
            "title: Authentication Cookie Intercepted\n"
            "logsource:\n  category: webserver\ndetection:\n"
            "  selection:\n    http_method: POST\n    uri|contains: '/login'\n  condition: selection"
        ),
    },
    "AitM-Proxy": {
        "ttp":      "T1557",
        "tactic":   "Collection",
        "name":     "Adversary-in-the-Middle",
        "splunk":   'index=proxy sourcetype=proxy\n| stats count by src_ip, dest_host\n| where count > 200',
        "sigma": (
            "title: Reverse Proxy / AitM Activity\n"
            "logsource:\n  category: proxy\ndetection:\n"
            "  selection:\n    http_method: POST\n  condition: selection"
        ),
    },
}

# ── Keyword fallbacks for modules not in the static map ──────────────────────

KEYWORD_FALLBACKS: list[tuple[str, str]] = [
    ("ssh",      "T1021.004"),
    ("smb",      "T1021.002"),
    ("ftp",      "T1071.002"),
    ("rdp",      "T1021.001"),
    ("http",     "T1190"),
    ("scan",     "T1046"),
    ("cred",     "T1552"),
    ("hash",     "T1003"),
    ("spray",    "T1110.003"),
    ("inject",   "T1055"),
    ("persist",  "T1547"),
    ("dns",      "T1071.004"),
    ("wifi",     "T1465"),
    ("payload",  "T1059"),
]


def _best_effort_ttp(module_name: str) -> dict:
    """Return a partial mapping dict for modules not in MITRE_MAPPINGS."""
    m = module_name.lower()
    for keyword, ttp in KEYWORD_FALLBACKS:
        if keyword in m:
            return {
                "ttp":    ttp,
                "tactic": "Unknown",
                "name":   f"Inferred from keyword '{keyword}'",
                "splunk": f'index=* | search "{module_name}"',
                "sigma":  f"title: {module_name} Activity\ndetection:\n  selection:\n    module: '{module_name}'\n  condition: selection",
            }
    return {
        "ttp":    "T0000",
        "tactic": "Unknown",
        "name":   "Unmapped Technique",
        "splunk": f'index=* | search module="{module_name}"',
        "sigma":  f"title: {module_name} Activity\ndetection:\n  condition: selection",
    }


# ─────────────────────────────────────────────────────────────────────────────
#  ENGINE
# ─────────────────────────────────────────────────────────────────────────────

class PurpleTeamEngine:

    # ── MITRE report ──────────────────────────────────────────────

    def generate_mitre_report(self) -> dict[str, dict]:
        """
        Build the ATT&CK table from the DB.
        Returns a dict of { module_name: mapping_dict } for modules that were seen.
        """
        console.print("[*] Analyzing Mission Database for MITRE ATT&CK mappings...")
        raw_logs = db.get_all()

        if not raw_logs:
            console.print(
                "[yellow][!] Database is empty. Execute modules first.[/yellow]"
            )
            return {}

        # Collect unique modules seen
        seen: dict[str, dict] = {}
        for row in raw_logs:
            mod = str(row.get("module") or "").strip()
            if not mod or mod in seen:
                continue
            mapping = MITRE_MAPPINGS.get(mod) or _best_effort_ttp(mod)
            seen[mod] = mapping

        if not seen:
            console.print("[yellow][!] No recognizable modules in the database.[/yellow]")
            return {}

        # Terminal table
        table = Table(
            title="Davoid Emulation → MITRE ATT&CK Matrix",
            border_style="bold magenta",
            expand=True,
        )
        table.add_column("Module Executed",  style="cyan")
        table.add_column("MITRE ID",         style="bold red")
        table.add_column("Tactic",           style="yellow")
        table.add_column("Technique Name",   style="white")

        for mod, m in seen.items():
            table.add_row(mod, m["ttp"], m.get("tactic", "—"), m["name"])

        console.print(table)
        return seen

    # ── Detection rules ───────────────────────────────────────────

    def generate_detection_rules(self, active_modules: dict[str, dict]):
        if not active_modules:
            return

        console.print(
            Panel("Generating Defensive Telemetry (Blue Team Artifacts)", border_style="bold blue")
        )

        for mod, mapping in active_modules.items():
            console.print(
                f"\n[bold green]=== {mapping['name']} ({mapping['ttp']}) ===[/bold green]"
            )

            # Splunk SPL with syntax highlighting
            console.print("[bold cyan]Splunk SPL:[/bold cyan]")
            console.print(Syntax(mapping["splunk"], "text", theme="monokai", line_numbers=False))

            # Sigma rule with YAML highlighting
            console.print("\n[bold yellow]Sigma Rule:[/bold yellow]")
            console.print(Syntax(mapping["sigma"], "yaml", theme="monokai", line_numbers=False))

            console.print("─" * 60)

    # ── Markdown export ───────────────────────────────────────────

    def export_markdown(self, active_modules: dict[str, dict]) -> str | None:
        """Write a Markdown report and return the filename."""
        if not active_modules:
            return None

        lines = [
            "# Davoid Purple Team Report",
            f"\n**Generated:** {datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n",
            "## MITRE ATT&CK Mapping\n",
            "| Module | MITRE ID | Tactic | Technique |",
            "|--------|----------|--------|-----------|",
        ]
        for mod, m in active_modules.items():
            lines.append(f"| {mod} | {m['ttp']} | {m.get('tactic','—')} | {m['name']} |")

        lines.append("\n## Detection Rules\n")
        for mod, m in active_modules.items():
            lines += [
                f"### {m['name']} ({m['ttp']})\n",
                "**Splunk SPL:**",
                f"```\n{m['splunk']}\n```\n",
                "**Sigma Rule:**",
                f"```yaml\n{m['sigma']}\n```\n",
                "---\n",
            ]

        fname = f"purple_team_{datetime.datetime.now().strftime('%Y%m%d_%H%M')}.md"
        try:
            with open(fname, "w", encoding="utf-8") as f:
                f.write("\n".join(lines))
            console.print(f"[bold green][+] Markdown report saved: {fname}[/bold green]")
            return fname
        except Exception as e:
            console.print(f"[red][!] Export failed: {e}[/red]")
            return None

    # ── ATT&CK Navigator export ───────────────────────────────────

    def export_navigator_layer(self, active_modules: dict[str, dict]) -> str | None:
        """
        Export an ATT&CK Navigator layer JSON.
        Import at https://mitre-attack.github.io/attack-navigator/
        """
        if not active_modules:
            return None

        techniques = []
        for mod, m in active_modules.items():
            ttp = m.get("ttp", "")
            if not ttp or ttp == "T0000":
                continue
            techniques.append({
                "techniqueID": ttp,
                "tactic":      m.get("tactic", "").lower().replace(" ", "-"),
                "color":       "#ff7b72",
                "comment":     f"Executed via {mod}",
                "enabled":     True,
                "score":       100,
            })

        layer = {
            "name":        "Davoid Campaign",
            "versions":    {"attack": "14", "navigator": "4.9", "layer": "4.5"},
            "domain":      "enterprise-attack",
            "description": f"Generated by Davoid on {datetime.datetime.now().isoformat()}",
            "techniques":  techniques,
            "gradient":    {"colors": ["#ff7b72", "#f0883e"], "minValue": 0, "maxValue": 100},
        }

        fname = f"navigator_layer_{datetime.datetime.now().strftime('%Y%m%d_%H%M')}.json"
        try:
            with open(fname, "w", encoding="utf-8") as f:
                json.dump(layer, f, indent=2)
            console.print(f"[bold green][+] ATT&CK Navigator layer saved: {fname}[/bold green]")
            console.print(
                "[dim]Import at: https://mitre-attack.github.io/attack-navigator/[/dim]"
            )
            return fname
        except Exception as e:
            console.print(f"[red][!] Navigator export failed: {e}[/red]")
            return None

    # ── Main menu ─────────────────────────────────────────────────

    def run(self):
        draw_header("Purple Team: Adversary Emulation & Detection")

        while True:
            choice = questionary.select(
                "Select Purple Team Operation:",
                choices=[
                    "1. Map Campaign to MITRE ATT&CK",
                    "2. Generate SIEM Detection Rules (Splunk / Sigma)",
                    "3. Export Markdown Report",
                    "4. Export ATT&CK Navigator Layer (JSON)",
                    "5. Full Run (Map + Rules + Export All)",
                    "Back",
                ],
                style=Q_STYLE,
            ).ask()

            if not choice or choice == "Back":
                break

            if "Map Campaign" in choice:
                self.generate_mitre_report()
                questionary.press_any_key_to_continue(style=Q_STYLE).ask()

            elif "SIEM" in choice:
                active = self.generate_mitre_report()
                self.generate_detection_rules(active)
                questionary.press_any_key_to_continue(style=Q_STYLE).ask()

            elif "Markdown" in choice:
                active = self.generate_mitre_report()
                self.export_markdown(active)
                questionary.press_any_key_to_continue(style=Q_STYLE).ask()

            elif "Navigator" in choice:
                active = self.generate_mitre_report()
                self.export_navigator_layer(active)
                questionary.press_any_key_to_continue(style=Q_STYLE).ask()

            elif "Full Run" in choice:
                active = self.generate_mitre_report()
                self.generate_detection_rules(active)
                self.export_markdown(active)
                self.export_navigator_layer(active)
                questionary.press_any_key_to_continue(style=Q_STYLE).ask()


def run_purple_team():
    PurpleTeamEngine().run()

File: internal/modules/purpleteam/purpleteam.go
package purpleteam

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/bryanparreira/davoid/internal/engagement"
	"github.com/bryanparreira/davoid/internal/modules/ui"
)

type technique struct {
	TID      string
	Name     string
	Tactic   string
	Splunk   string
	Sigma    string
	Modules  []string
}

var mitreTTPs = []technique{
	{
		TID:    "T1595.001",
		Name:   "Active Scanning: Scanning IP Blocks",
		Tactic: "Reconnaissance",
		Splunk: `index=network sourcetype=firewall | stats count by src_ip | where count > 100`,
		Sigma:  "detection:\n  keywords:\n    - 'nmap'\n    - 'masscan'",
		Modules: []string{"scanner"},
	},
	{
		TID:    "T1596.005",
		Name:   "Search Open Technical Databases",
		Tactic: "Reconnaissance",
		Splunk: `index=dns | stats count by query | sort -count`,
		Sigma:  "detection:\n  keywords:\n    - 'subdomain enumeration'",
		Modules: []string{"osint"},
	},
	{
		TID:    "T1592",
		Name:   "Gather Victim Host Information",
		Tactic: "Reconnaissance",
		Splunk: `index=web | stats count by user_agent | sort -count`,
		Sigma:  "detection:\n  keywords:\n    - 'user-agent scanner'",
		Modules: []string{"web_recon", "osint"},
	},
	{
		TID:    "T1557.002",
		Name:   "ARP Cache Poisoning",
		Tactic: "Collection",
		Splunk: `index=network sourcetype=arp | stats count by src_mac dest_mac | where src_mac != dest_mac`,
		Sigma:  "detection:\n  keywords:\n    - 'arp-reply'\n    - 'gratuitous arp'",
		Modules: []string{"mitm"},
	},
	{
		TID:    "T1040",
		Name:   "Network Sniffing",
		Tactic: "Credential Access",
		Splunk: `index=network sourcetype=zeek_conn | stats count by proto | sort -count`,
		Sigma:  "detection:\n  keywords:\n    - 'promiscuous mode'\n    - 'tcpdump'",
		Modules: []string{"sniff"},
	},
	{
		TID:    "T1566.002",
		Name:   "Phishing: Spearphishing Link",
		Tactic: "Initial Access",
		Splunk: `index=web | where uri="/login" AND method="POST" | stats count by src_ip`,
		Sigma:  "detection:\n  keywords:\n    - 'credential harvest'\n    - 'clone site'",
		Modules: []string{"phishing"},
	},
	{
		TID:    "T1105",
		Name:   "Ingress Tool Transfer",
		Tactic: "Command and Control",
		Splunk: `index=network | where bytes_out > 1000000 | stats sum(bytes_out) by dest_ip`,
		Sigma:  "detection:\n  keywords:\n    - 'wget'\n    - 'curl'\n    - 'certutil'",
		Modules: []string{"payloads"},
	},
	{
		TID:    "T1071.001",
		Name:   "C2: Web Protocols",
		Tactic: "Command and Control",
		Splunk: `index=proxy | where bytes < 1000 AND status=200 | bucket _time span=5m | stats count by dest | where count > 50`,
		Sigma:  "detection:\n  keywords:\n    - 'beacon'\n    - 'c2 callback'\n    - 'implant'",
		Modules: []string{"ghost_hub"},
	},
	{
		TID:    "T1027",
		Name:   "Obfuscated Files or Information",
		Tactic: "Defense Evasion",
		Splunk: `index=endpoint | where process_name="python*" AND cmdline="*base64*"`,
		Sigma:  "detection:\n  keywords:\n    - 'base64'\n    - 'encrypted payload'",
		Modules: []string{"crypt_keeper", "payloads"},
	},
	{
		TID:    "T1547.001",
		Name:   "Boot or Logon Autostart: Registry Run Keys",
		Tactic: "Persistence",
		Splunk: `index=windows source="WinEventLog:Security" EventCode=13 | where TargetObject LIKE "%Run%"`,
		Sigma:  "detection:\n  keywords:\n    - 'HKCU\\\\Run'\n    - 'HKLM\\\\Run'",
		Modules: []string{"persistence"},
	},
	{
		TID:    "T1110.003",
		Name:   "Password Spraying",
		Tactic: "Credential Access",
		Splunk: `index=windows source="WinEventLog:Security" EventCode=4625 | stats count by src_ip | where count > 10`,
		Sigma:  "detection:\n  keywords:\n    - 'authentication failure'\n    - 'spray'",
		Modules: []string{"cred_tester", "ad_ops"},
	},
	{
		TID:    "T1003.001",
		Name:   "OS Credential Dumping: LSASS",
		Tactic: "Credential Access",
		Splunk: `index=windows source="WinEventLog:Security" EventCode=10 TargetImage="*lsass.exe"`,
		Sigma:  "detection:\n  keywords:\n    - 'mimikatz'\n    - 'lsass'",
		Modules: []string{"looter"},
	},
	{
		TID:    "T1018",
		Name:   "Remote System Discovery",
		Tactic: "Discovery",
		Splunk: `index=network sourcetype=firewall | stats count by src_ip dest_port | where count > 50`,
		Sigma:  "detection:\n  keywords:\n    - 'port scan'\n    - 'host discovery'",
		Modules: []string{"scanner"},
	},
	{
		TID:    "T1021.004",
		Name:   "Remote Services: SSH",
		Tactic: "Lateral Movement",
		Splunk: `index=linux source="/var/log/auth.log" "Accepted password"`,
		Sigma:  "detection:\n  keywords:\n    - 'ssh login'\n    - 'paramiko'",
		Modules: []string{"looter", "cred_tester"},
	},
	{
		TID:    "T1078.002",
		Name:   "Valid Accounts: Domain Accounts",
		Tactic: "Defense Evasion",
		Splunk: `index=windows source="WinEventLog:Security" EventCode=4624 LogonType=3`,
		Sigma:  "detection:\n  keywords:\n    - 'domain login'\n    - 'ldap bind'",
		Modules: []string{"ad_ops"},
	},
}

func Run() error {
	ui.Header("Purple Team — MITRE ATT&CK Mapper & Detection Engineer")

	eng, _ := engagement.Active()
	if eng == nil {
		ui.Warn("No active engagement — showing all TTPs.")
	}

	for {
		action := ui.Select("Purple Team Operation", []string{
			"View MITRE ATT&CK Coverage",
			"Generate Detection Rules (Splunk SPL)",
			"Generate Sigma Rules",
			"Export ATT&CK Navigator JSON",
			"Generate Markdown Report",
		})
		if action < 0 {
			break
		}

		switch action {
		case 0:
			showCoverage(eng)
		case 1:
			generateSplunk(eng)
		case 2:
			generateSigma(eng)
		case 3:
			exportNavigator()
		case 4:
			generateMarkdown(eng)
		}
	}
	return nil
}

func showCoverage(_ *engagement.Engagement) {
	fmt.Println()
	ui.Divider()
	fmt.Printf("  %-12s  %-15s  %-40s  %s\n",
		ui.Bold.Render("TID"),
		ui.Bold.Render("TACTIC"),
		ui.Bold.Render("TECHNIQUE"),
		ui.Bold.Render("MODULES"),
	)
	ui.Divider()

	tactics := map[string][]technique{}
	for _, t := range mitreTTPs {
		tactics[t.Tactic] = append(tactics[t.Tactic], t)
	}

	tacticOrder := []string{
		"Reconnaissance", "Initial Access", "Execution", "Persistence",
		"Defense Evasion", "Credential Access", "Discovery",
		"Lateral Movement", "Collection", "Command and Control", "Exfiltration",
	}

	for _, tactic := range tacticOrder {
		ttps, ok := tactics[tactic]
		if !ok {
			continue
		}
		fmt.Println()
		fmt.Println(ui.Yellow.Render("  ── " + tactic + " ──"))
		for _, t := range ttps {
			fmt.Printf("  %-12s  %-15s  %-40s  %s\n",
				ui.Cyan.Render(t.TID),
				"",
				t.Name,
				ui.Dim.Render(strings.Join(t.Modules, ", ")),
			)
		}
	}
	fmt.Printf("\n  Total: %d techniques mapped.\n", len(mitreTTPs))
	ui.PressEnter()
}

func generateSplunk(_ *engagement.Engagement) {
	fmt.Println()
	ui.Divider()
	ui.Info("Splunk SPL Detection Rules")
	ui.Divider()
	for _, t := range mitreTTPs {
		fmt.Printf("\n  %s  %s\n", ui.Cyan.Render(t.TID), t.Name)
		fmt.Println(ui.Yellow.Render("  " + t.Splunk))
	}
	ui.PressEnter()
}

func generateSigma(_ *engagement.Engagement) {
	fmt.Println()
	ui.Divider()
	ui.Info("Sigma Detection Rules")
	ui.Divider()
	for _, t := range mitreTTPs {
		fmt.Printf("\n  %s  %s\n", ui.Cyan.Render(t.TID), t.Name)
		rule := fmt.Sprintf("title: Detect %s\nstatus: experimental\ndescription: Detects %s\ntags:\n  - attack.%s\n  - %s\n%s",
			t.Name, t.Name, strings.ToLower(strings.ReplaceAll(t.Tactic, " ", "_")), strings.ToLower(t.TID), t.Sigma)
		for _, line := range strings.Split(rule, "\n") {
			fmt.Println(ui.Dim.Render("  " + line))
		}
	}
	ui.PressEnter()
}

type navigatorLayer struct {
	Name        string        `json:"name"`
	Version     string        `json:"version"`
	Domain      string        `json:"domain"`
	Description string        `json:"description"`
	Techniques  []navTechnique `json:"techniques"`
}

type navTechnique struct {
	TechniqueID string `json:"techniqueID"`
	Score       int    `json:"score"`
	Comment     string `json:"comment"`
}

func exportNavigator() {
	layer := navigatorLayer{
		Name:        "Davoid Coverage",
		Version:     "4.4",
		Domain:      "enterprise-attack",
		Description: "Davoid red team framework ATT&CK coverage",
	}
	for _, t := range mitreTTPs {
		layer.Techniques = append(layer.Techniques, navTechnique{
			TechniqueID: t.TID,
			Score:       75,
			Comment:     "Covered by: " + strings.Join(t.Modules, ", "),
		})
	}
	data, _ := json.MarshalIndent(layer, "", "  ")
	fname := fmt.Sprintf("attack_navigator_%d.json", time.Now().Unix())
	os.WriteFile(fname, data, 0644)
	ui.Success(fmt.Sprintf("ATT&CK Navigator layer exported: %s", fname))
	ui.PressEnter()
}

func generateMarkdown(eng *engagement.Engagement) {
	var sb strings.Builder
	sb.WriteString("# Purple Team Report — MITRE ATT&CK Coverage\n\n")
	if eng != nil {
		sb.WriteString(fmt.Sprintf("**Engagement:** %s\n\n", eng.Name))
	}
	sb.WriteString(fmt.Sprintf("**Generated:** %s\n\n", time.Now().Format(time.RFC1123)))
	sb.WriteString("## Technique Coverage\n\n")
	sb.WriteString("| TID | Tactic | Technique | Detection |\n")
	sb.WriteString("|-----|--------|-----------|----------|\n")
	for _, t := range mitreTTPs {
		sb.WriteString(fmt.Sprintf("| %s | %s | %s | [Splunk] |\n", t.TID, t.Tactic, t.Name))
	}

	fname := fmt.Sprintf("purple_team_report_%d.md", time.Now().Unix())
	os.WriteFile(fname, []byte(sb.String()), 0644)
	ui.Success(fmt.Sprintf("Purple team report: %s", fname))
	ui.PressEnter()
}
